swap/src/protocol/alice/spot_price.rs

```diff
@@ -0,0 +1,201 @@
+                vec![(SpotPriceProtocol, ProtocolSupport::Inbound)],
+                RequestResponseConfig::default(),
+            ),
+            events: Default::default(),
+            balance,
+            lock_fee,
+            max_buy,
+            latest_rate,
+            resume_only,
+        }
+    }
+
+    pub fn update_balance(&mut self, balance: monero::Amount) {
+        self.balance = balance;
+    }
+
+    fn handle_error_response(
+        &mut self,
+        peer: PeerId,
+        channel: ResponseChannel<spot_price::Response>,
+        error: spot_price::Error,
+    ) {
+        if self
+            .behaviour
+            .send_response(channel, spot_price::Response::Error(error))
+            .is_err()
+        {
+            tracing::debug!(%peer, "Unable to send error response for spot price request");
+        }
+    }
+
+    fn poll<BIE>(
+        &mut self,
+        _cx: &mut Context<'_>,
+        _params: &mut impl PollParameters,
+    ) -> Poll<NetworkBehaviourAction<BIE, OutEvent>> {
+        if let Some(event) = self.events.pop_front() {
+            return Poll::Ready(NetworkBehaviourAction::GenerateEvent(event));
+        }
+
+        // We trust in libp2p to poll us.
+        Poll::Pending
+    }
+}
+
+impl<LR> NetworkBehaviourEventProcess<spot_price::OutEvent> for Behaviour<LR>
+where
+    LR: LatestRate + Send + 'static,
+{
+    fn inject_event(&mut self, event: spot_price::OutEvent) {
+        let (peer, message) = match event {
+            RequestResponseEvent::Message { peer, message } => (peer, message),
+            RequestResponseEvent::OutboundFailure { peer, error, .. } => {
+                tracing::error!(%peer, "Failure sending spot price response: {:#}", error);
+                return;
+            }
+            RequestResponseEvent::InboundFailure { peer, error, .. } => {
+                tracing::warn!(%peer, "Inbound failure when handling spot price request: {:#}", error);
+                return;
+            }
+            RequestResponseEvent::ResponseSent { peer, .. } => {
+                tracing::debug!(%peer, "Spot price response sent");
+                return;
+            }
+        };
+
+        let (request, channel) = match message {
+            RequestResponseMessage::Request {
+                request, channel, ..
+            } => (request, channel),
+            RequestResponseMessage::Response { .. } => {
+                tracing::error!("Unexpected message");
+                return;
+            }
+        };
+
+        if self.resume_only {
+            tracing::warn!(%peer, "Ignoring spot price request from {} because ASB is running in maintenance mode", peer);
+            self.handle_error_response(peer, channel, spot_price::Error::MaintenanceMode);
+            return;
+        }
+
+        let btc = request.btc;
+        if btc > self.max_buy {
+            tracing::warn!(%peer, "Ignoring spot price request from {} because max muy amount exceeded", peer);
```

Comment:
If you attach this log message to the error variant, you could have a generic log inside the function.

Reply:
I thought about that but could not come up with an easy way to use the `peer` field as context and logging on different levels. Am I missing something, can this be achieved in an easy way?
I did look into the `tracing::Event` and potentially passing events around and then logging them, but I think that makes our life harder than it has to be.

Reply:
> I thought about that but could not come up with an easy way to use the peer field as context and logging on different levels if I move it in. Am I missing something, can this be achieved in an easy way?

You always have a peer right? You can pass it as a standalone parameter together with the error and then always supply it as context and simply log the error with `{:#}`.

Reply:
But I would not be able to adapt the level... some logs are `warn` and other are `error` or am I missing something?